### ISSUE TITLE
support training instance segmentation model directly from masks

### ISSUE THREAD

**Issue by stevensagaert** [ultralytics/ultralytics#17882]
### Search before asking

- [X] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Currently only polygon annotations are supported and the yolo format for it is rather restrictive (one contiguous contour per mask/instance). So no support for multiple disjoint parts or masks with holes in them.

Also all the conversion scripts I have tried (even wrote a few myself) from mask data to polygon always have some issues with complicated cases. There are a a lot of edge cases where it will fail.  

Mask data is just a lot more accurate and there is no information loss when trying to solve the problem mask -> contour. So I'd like support for a common mask annotation format like COCO.

### Use case

training instance segmentation model with complex masks or a lot of masks (whole image segmented)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @stevensagaert, thank you for your feature request and for contributing to the Ultralytics community 🚀! 

We understand your concern regarding mask-to-polygon conversion and its potential limitations when dealing with complex or disjoint masks. Supporting direct training from mask annotations like COCO format would indeed provide a more seamless and accurate workflow for segmentation tasks. Your suggestion is much appreciated and will be reviewed carefully by the Ultralytics team 💡.

If you'd like to elaborate on your request further or provide additional context, feel free to do so! If this feature is something you'd like to contribute to directly, let us know—we're always open to collaborating on Pull Requests (PRs) 🙌.

To ensure your setup is optimal, we recommend verifying you're using the latest version of the `ultralytics` package. You can upgrade with:

```bash
pip install -U ultralytics
```

### Resources
For related details, you may find the following resources helpful:
- [Docs](https://docs.ultralytics.com) contain guides on [Tasks](https://docs.ultralytics.com/tasks/) and [Model Training Tips](https://docs.ultralytics.com/guides/model-training-tips/).
- Real-time community discussions are available on [Discord](https://discord.com/invite/ultralytics) 🎧. For other forms of collaboration, check out [Discourse](https://community.ultralytics.com) or [Reddit](https://reddit.com/r/Ultralytics).

### Environment Recommendations
Run YOLO in any verified environment with all dependencies, including [Python>=3.8](https://www.python.org/) and [PyTorch>=1.8](https://pytorch.org/), pre-installed:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Cloud Computing**: [GCP](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) or [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/).
- **Docker Image**: Use our [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>.

### Status
By the way, you can check the health of the repository's continuous integration tests here:  
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>. Green status indicates all is running smoothly.

This is an automated response designed to guide you quickly to helpful resources and best practices. An Ultralytics engineer will review your request and provide further assistance soon. Thank you again for your input and ongoing support 🌟!

**Comment by Y-T-G** [comment]
I don't think COCO supports PNG masks either (which is why they have the limitation as YOLO format that masks can't have holes https://github.com/cocodataset/cocoapi/issues/153). PNG masks are usually used for semantic segmentation. Instance segmentation usually uses polygons.

MMDet also doesn't support PNG masks for instance segmentation. Neither does PaddleDetection.

MMSeg does, but it's because it's a framework for semantic segmentation.

**Comment by stevensagaert** [comment]
In the coco annotation format you can have either polygon annotations, but you can also have mask RLE encoded. So annotation tools (segments.ai) export in that format.
________________________________

**Comment by stevensagaert** [comment]
A tool like segments.ai by default also exports their annotations as binary images.
________________________________

**Comment by stevensagaert** [comment]
Yes I have seen that issue. Too bad augmentations are killing that but I'm not surprised. Can' t pytorch  TVtensors not deal with that? Or Albumentations?
________________________________

**Comment by Y-T-G** [comment]
With masks, I guess you will have to mimic every transformation that is applied to the image on the masks too in parallel. Currently, the augmentations all work with polygons, so there are no mask counterparts.

Ultralytics uses custom transforms (mosaic, mixup, copy paste, random perspective), so Albumentations wouldn't help much.
